### PR TITLE
Use integer index instead of string for object identifier

### PR DESCRIPTION
**Pull request by junha1**

### CODE REVIEW

File: remote-trait-object/src/service.rs [hunk 1]
```diff
@@ -14,15 +14,21 @@
...
 #[derive(PartialEq, Serialize, Deserialize, Debug, Clone, Copy)]
-pub struct ServiceObjectId {
-    pub(crate) index: InstanceId,
+pub struct HandleToExchange(pub(crate) ServiceObjectId);
+
+impl HandleToExchange {
+    /// FIXME
+    /// This function is for incomplete tests' temporary remote object creations.
+    /// You should remove this in the future.
+    pub fn debug_singleton() -> Self {
```
Comment (majecty) [inline remote-trait-object/src/service.rs:29]: This function is not used.

File: remote-trait-object/src/forwarder.rs [hunk 2]
```diff
@@ -37,24 +39,30 @@ impl ServiceForwarder {
...
+            }),
         }
     }
 
-    pub fn add_service(&self, name: ServiceId, service: Box<dyn Dispatch>) {
-        let insert_result = self.service_handlers.write().insert(name.clone(), service);
-        if insert_result.is_some() {
-            panic!("Duplicated service id {}", name);
-        }
+    pub fn add_service(&self, service: Arc<dyn Dispatch>) -> ServiceObjectId {
```
Comment (majecty) [inline remote-trait-object/src/forwarder.rs:52]: How about `add_service_object(&self, service_object: Arc<dyn Dispatch>)`?

File: remote-trait-object/src/packet.rs [hunk 3]
```diff
@@ -80,8 +81,7 @@ const SLOT_CALL_OR_RETURN_INDICATOR: SlotId = SlotId(1000);
 #[repr(C)]
 struct PacketHeader {
     pub slot: SlotId,
-    // FIXME: use integer indices for the service
-    pub target_service_name: [u8; 100],
+    pub object: ServiceObjectId,
```
Comment (majecty) [inline remote-trait-object/src/packet.rs:84]: How about `service_object` or `service_object_id`?
`Object` is a too common name.

File: remote-trait-object/src/forwarder.rs [hunk 4]
```diff
@@ -18,13 +18,15 @@ use crate::packet::PacketView;
...
+use std::collections::{HashMap, VecDeque};
 use std::fmt;
+use std::sync::Arc;
 
-type ServiceId = String;
+pub type ServiceObjectId = u32;
 
 pub struct ServiceForwarder {
-    service_handlers: RwLock<HashMap<ServiceId, Box<dyn Dispatch>>>,
+    service_handlers: RwLock<HashMap<ServiceObjectId, Arc<dyn Dispatch>>>,
```
Comment (majecty) [inline remote-trait-object/src/forwarder.rs:28]: How about `service_objects`?
```suggestion
    service_objects: RwLock<HashMap<ServiceObjectId, Arc<dyn Dispatch>>>,
```

File: remote-trait-object/src/forwarder.rs [hunk 5]
```diff
@@ -18,13 +18,15 @@ use crate::packet::PacketView;
...
+    tokens: RwLock<VecDeque<ServiceObjectId>>,
```
Comment (majecty) [inline remote-trait-object/src/forwarder.rs:29]: How about `available_ids` or `available_service_object_ids`?

File: remote-trait-object/src/port.rs [hunk 6]
```diff
@@ -28,8 +29,7 @@ use std::sync::Arc;
 pub trait Port: Send + Sync + 'static {
     fn call(&self, packet: PacketView) -> Packet;
     fn delete_request(&self, id: ServiceObjectId);
-    /// TODO: Assign id automatically and return it.
-    fn register(&self, id: String, handle_to_register: Box<dyn Dispatch>);
+    fn register(&self, service: Arc<dyn Dispatch>) -> HandleToExchange;
```
Comment (majecty) [inline remote-trait-object/src/port.rs:32]: How about service_object?
```suggestion
    fn register(&self, service_object: Arc<dyn Dispatch>) -> HandleToExchange;
```

File: remote-trait-object/src/forwarder.rs (hunk 2 above)
Comment (junha1) [reply remote-trait-object/src/forwarder.rs]: @majecty I used `register` instead of `add`.

File: remote-trait-object/src/forwarder.rs [hunk 7]
```diff
@@ -18,43 +18,51 @@ use crate::packet::PacketView;
...
+            }),
         }
     }
 
-    pub fn add_service(&self, name: ServiceId, service: Box<dyn Dispatch>) {
-        let insert_result = self.service_handlers.write().insert(name.clone(), service);
-        if insert_result.is_some() {
-            panic!("Duplicated service id {}", name);
-        }
+    pub fn register_service_object(&self, service: Arc<dyn Dispatch>) -> ServiceObjectId {
```
Comment (majecty) [inline remote-trait-object/src/forwarder.rs:52]: How about service_object?
```suggestion
    pub fn register_service_object(&self, service_object: Arc<dyn Dispatch>) -> ServiceObjectId {
```

File: remote-trait-object/src/forwarder.rs [hunk 8]
```diff
@@ -18,43 +18,51 @@ use crate::packet::PacketView;
...
+        let token = self.available_ids.write().pop_front().expect("Too many service objects had been created");
```
Comment (majecty) [inline remote-trait-object/src/forwarder.rs:53]: IMO, we don't need use the name `token`. It does not have any meaning. 
```suggestion
        let service_object_id = self.available_ids.write().pop_front().expect("Too many service objects had been created");
```

File: remote-trait-object/src/packet.rs [hunk 9]
```diff
@@ -200,9 +190,9 @@ impl Packet {
         packet
     }
 
-    pub fn new_request(service_name: String, method: MethodId, args: &[u8]) -> Self {
+    pub fn new_request(object: ServiceObjectId, method: MethodId, args: &[u8]) -> Self {
```
Comment (majecty) [inline remote-trait-object/src/packet.rs:193]: `object` is too general. How about `object_id` or `service_object_id`?

```suggestion
    pub fn new_request(service_object_id: ServiceObjectId, method: MethodId, args: &[u8]) -> Self {
```

**Pull request merged**